lib/buffer.js

```diff
@@ -272,19 +272,66 @@ Buffer.concat = function(list, length) {
 };
 
 
+function base64ByteLength(str) {
+  var len = str.length;
+  var bytes = len;
+
+  // Handle padding
+  if (str[len - 1] === '=')
+    bytes--;
+  if (len > 2 && str[len - 2] === '=')
+    bytes--;
+
+  // Base64 ratio: 3/4
+  bytes = (bytes / 4) * 3;
+
+  return Math.floor(bytes);
+}
+
+function utf8ByteLength(str) {
+  var bytes = str.length;
+
+  for (var i = 0, l = str.length; i < l; i++) {
+    var code = str.charCodeAt(i);
+
+    // Based on where the code lies (0x7F-0x10FFFF), derive char value.
+    // One byte for each character is already covered by str.length, so
+    // only 1 must be added for 2, 2 for 3, etc.
+    if (code <= 0x7F)
+      { } // nop
+    else if (code <= 0x7FF)
+      bytes += 1;
+    else if (code <= 0xFFFF)
+      bytes += 2;
+    else if (code <= 0x10FFFF)
+      bytes += 3;
+  }
+
+  return bytes;
+}
+
+
 function byteLength(string, encoding) {
-  if (typeof(string) !== 'string')
-    string = String(string);
+  if (typeof string !== 'string')
+    string = '' + string;
+  if (typeof encoding !== 'string')
+    encoding = 'utf8';
 
   if (string.length === 0)
     return 0;
 
-  switch (encoding) {
+  switch (encoding.toLowerCase()) {
```

Comment:
`String.prototype.toLowerCase()` is pretty expensive, see 4ddd640 for an example.